hal/src/rtl872x/i2c_hal.cpp

```diff
@@ -474,19 +507,60 @@ class I2cClass {
     }
 
 private:
+    enum I2cSlaveStatus {
+        I2C_SLAVE_STOPPED,
+        I2C_SLAVE_STARTED,
+        I2C_SLAVE_RESTARTED,
+        I2C_SLAVE_TX,
+        I2C_SLAVE_RX,
+    };
+
     I2cClass(I2C_TypeDef* i2cDev, hal_pin_t sda, hal_pin_t scl)
             : i2cDev_(i2cDev),
               sdaPin_(sda),
               sclPin_(scl),
               configured_(false),
+              slaveStatus_(I2C_SLAVE_STOPPED),
               heapBuffer_(false),
               i2cInitStruct_(),
               onRequested_(nullptr),
               onReceived_(nullptr),
-              mutex_(nullptr) {
+              mutex_(nullptr),
+              slaveRxCache_(nullptr),
+              slaveRxCacheLen_(0),
+              slaveRxCacheDepth_(HAL_PLATFORM_I2C_BUFFER_SIZE(HAL_I2C_INTERFACE1) * 2) {
     }
     ~I2cClass() = default;
 
+    bool masterRestarted() {
+        return (i2cDev_->IC_RAW_INTR_STAT & BIT_IC_INTR_STAT_R_START_DET) &&
+                !(i2cDev_->IC_RAW_INTR_STAT & BIT_IC_INTR_STAT_R_STOP_DET);
+                (i2cDev_->IC_RAW_INTR_STAT & BIT_IC_INTR_STAT_R_ACTIVITY);
```

Comment:
Does this line do anything? 
Is the `;` at the end of line 537 intentional? Should it be an extra condition ie `&&`?

Reply:
Ah, good catch! It should be `&&`.